### DISCUSSION TITLE
Tools, dependencies, and other things.

### DISCUSSION THREAD

**Discussion by rquadling** [opentofu/opentofu#1104] [General]
Hello.

As Terraform is a well established product, with lots of tools all working on the back of it, knowing what they all are, and if they are being updated to take into account OpenToFu, well that's useful.

Please add the tools, plugins, utilities that are NOT part of Terraform's direct offering and see where they are in terms of supporting with OpenToFu. Maybe some of these become part of OpenToFu's standard offering rather than being an external tool!

If you add them, it would be useful to have 1 comment per tool and any information you know about with regard to the tools support for OpenToFu.

1. [`pre-commit-terraform`](https://github.com/orgs/opentofu/discussions/1104#discussioncomment-8096836)
2. [`tflint`](https://github.com/orgs/opentofu/discussions/1104#discussioncomment-8098010)
3. [`tfenv`](https://github.com/orgs/opentofu/discussions/1104#discussioncomment-8098016)
4. [`terraform-docs`](https://github.com/orgs/opentofu/discussions/1104#discussioncomment-8098018)
5. [`asdf`](https://github.com/orgs/opentofu/discussions/1104#discussioncomment-8099843) (Thank you [David Gamba](https://github.com/DavidGamba))
6. [`trivy`](https://github.com/orgs/opentofu/discussions/1104#discussioncomment-8105890) (Thank you [David J Eddy](https://github.com/DavidGamba))
7. [`kics`](https://github.com/orgs/opentofu/discussions/1104#discussioncomment-8105890) (Thank you [David J Eddy](https://github.com/DavidGamba))
8. [`tfsec`](https://github.com/orgs/opentofu/discussions/1104#discussioncomment-8105890) (Thank you [David J Eddy](https://github.com/DavidGamba))

I've kept each of the tools as a separate comment, so replies related can be kept together.

I'm not sure anyone can edit this, but if you add a new tool, I'll update the above list as soon as I'm able to.

Thank you!

**Comment by rquadling** [comment]
# `pre-commit-terraform`
The [`pre-commit-terraform`](https://github.com/antonbabenko/pre-commit-terraform) plugin for [`pre-commit`](https://pre-commit.com/) is used a lot for the projects I'm involved in. Also not got anything related to OpenToFu at the moment.

**Reply by kvendingoldo** [reply]
Hi! Out team forked the project and plan to adopt it for OpenTofu by the end of this week.
Welcome to contribute: https://github.com/tofuutils/pre-commit-opentofu

**Comment by rquadling** [comment]
# `tflint`
The [`tflint`](https://github.com/terraform-linters/tflint) tool has got a [proposal](https://github.com/terraform-linters/tflint/issues/1878) and [discussion](https://github.com/terraform-linters/tflint/discussions/1826) regarding OpenToFu. Will be worth seeing how that goes.

**Reply by Mythra** [reply]
Following up here, curious to see what other folks have been doing. Seems the discussion is going to be focused on the HCL Language spec (but will also become BUSL on binary distributions). Have folks seen tflint continue to work on opentofu, do we expect it to continue to work, any forks of the tool popped up yet?

**Reply by iamleot** [reply]
Just for the record, `tflint` version [0.51.0 and newer updated embedded newer Terraform package and so it is affected by BUSL license](https://github.com/terraform-linters/tflint/releases/tag/v0.51.0).

Last `tflint` version not affected is version 0.50.3.

(Leaving it here only for completeness.)

**Reply by alexjfisher** [reply]
I've just discovered I can't use `tflint` in one of my projects as it uses a variable in a module `source`.  Something that `opentofu` supports, but apparently terraform doesn't...

```
Failed to load configurations; main.tf:53,14-17: Variables not allowed; Variables may not be used here., and 1 other diagnostic(s):
Error: Variables not allowed
  on main.tf line 53, in module "alb":
  53:   source   = var.alb_module_source
Variables may not be used here.
```

**Reply by zhimsel** [reply]
@alexjfisher I'm facing the same issue. That's probably the biggest hurdle with using some of the newer OpenTofu features (like expressions in module sources). I don't want to lose linting entirely to support that feature in my codebase.

**Reply by kvendingoldo** [reply]
Some time ago, me and tofuutils team were thinking about [tflint fork](https://github.com/tofuutils/tofulint), but it was useless at that time. How do you think is it a time to restore this idea?

**Reply by zhimsel** [reply]
> do you think is it a time to restore this idea?

I'd definitely argue that it is. There's enough differences at this point to warrant a fork. Module source expressions is one example. State encryption is another.

**Reply by git-ival** [reply]
Has there been any traction on forking `tflint` to have opentofu support? As `tflint` doesn't support opentofu, it doesn't support features that terraform does not support. For example,`tflint` does not support variables in module source fields which is an often-used feature for larger projects. Instead, `tflint` treats this as an error.

**Reply by creeder-uturn** [reply]
@kvendingoldo I would encourage it! I've loved the work ya'll are doing with tenv, the same for tflint could be great.

**Reply by ImIOImI** [reply]
Unless something has changed, it appears that tflint [is not planning on implementing tofu support](https://github.com/terraform-linters/tflint/issues/2194) so a fork may be the only viable option.

**Reply by kvendingoldo** [reply]
We had it in our plans and have already started to look at an ability to make a fork (https://github.com/tofuutils/tofulint) that will work for Terraform and OpenTofu together, but we need to have more hands for that. We'll be glad to welcome anyone who wants to be involved.

**Reply by zhimsel** [reply]
> work for Terraform and OpenTofu together

If it's a true fork, it makes sense to not bother supporting Terraform. The two codebases are already starting to diverge and there _may_ be legal complications when dealing with post 1.5  Terraform support.

It would also reduce the maintenance/development overhead to only aim for Tofu support.

**Reply by joeaawad** [reply]
As Alex said, the lack of a linting option is also blocking my company from implementing any OpenTofu specific syntax features like variables in module source. On my end, we do not have any preference if a linter is implemented as a stand alone tool like a fork of tflint or included in tofu itself, as https://github.com/opentofu/opentofu/issues/2213 mentions. 

The thing that we do have a strong preference on is that the solution should either be backwards compatible with existing custom tflint rules or allow for easy refactoring of custom tflint rules to implement them in this new tool with minimal effort since we do have fairly extensive custom lint rules.

**Reply by dragonfleas** [reply]
I thought about mentioning this inside my tofu lint issue, but since it's more discussion-oriented, I'll put it here. This is more of a meta-analysis and justification for such a tool anyway.

Right now, as the landscape is changing at such a rapid pace due to AI-generated code, it's become **paramount** to achieve fast feedback loops back to AI that assert the guardrails of specifications your organization puts in place. At my current organization, we're fully committing to _smart_ generative AI use, which means establishing strict adherence to our linting rules, testing practices, and dependency tree constraints set forth by our tools in various coding ecosystems.

As the sole DevOps engineer at my organization, it's hard for me to establish a use case for tools like Claude Code (this extends to any generative AI coding tool) because the tooling in the TF ecosystem is so immature that we cannot feel confident that the changes introduced will adhere to the guidelines, regardless of whether they are in prompt context or not. Moving to increased productivity requires a prerequisite and a heightened ceiling for strong architectural guidance and decision-making that can be taught to engineers, but needs to be _enforced_ and _asserted_ against the AI.

Ideally we have something sooner rather than later, so if a `tflint` fork would be faster to integrate than adopting first-party support for an "in-engine" linting tool, I vote for that.

**Reply by iamleot** [reply]
@diofeher I think so. As pointed out in the discussion there are possible new (OpenTofu) language differences, e.g. https://github.com/orgs/opentofu/discussions/1104#discussioncomment-12316023, that are not supported by tflint 0.50.3 that is the last non-BUSL version and errors are actually from tflint itself, not part of https://github.com/terraform-linters/tflint-ruleset-terraform.

**Reply by diofeher** [reply]
@iamleot Thanks for answering. I previously deleted my comment because I spent some time looking and responded to my own question!

But for others, I was asking whether just forking https://github.com/terraform-linters/tflint-ruleset-terraform would be enough, or if a fork of https://github.com/terraform-linters/tflint was needed as well.

**Reply by zhimsel** [reply]
> whether just forking https://github.com/terraform-linters/tflint-ruleset-terraform would be enough, or if a fork of https://github.com/terraform-linters/tflint was needed as well.

TFLint itself would need to be forked. A number of the Tofu-specific features cause panics/exceptions/etc during TFLint's parsing of the HCL itself. For instance, it's not expecting a `for_each` to be in a `provider` block, and balks at that. It doesn't even _get_ to rule evaluation 😄

**Reply by kvendingoldo** [reply]
Absolutely agree with u. Tofuutils team tried to do that, but we don't have enough capacity for that. If you'd like to join we can restore [the project ](https://github.com/tofuutils/tofulint).

**Comment by rquadling** [comment]
# `tfenv`
I've used [`tfenv`](https://github.com/tfutils/tfenv) for the projects I'm involved in. But the upgrade for it to be opentofu compatible were [rejected](https://github.com/tfutils/tfenv/issues/409#issuecomment-1863021455). Luckily [`tofuenv`](https://github.com/tofuutils/tofuenv) exists!

**Reply by dvaumoron** [reply]
I have done a go rewrite called gotofuenv, however we are discussing to see if my work can be embedded in their tenv project (currently a go wrapper calling tfenv and tofuenv but they want to rewrite it as a go standalone)

**Reply by kvendingoldo** [reply]
Yep. [tofuenv](https://github.com/tofuutils/tofuenv) is a good tool, but it's legacy. Some time ago we started to work on [tenv](https://github.com/tofuutils/tenv) that will be golang native application for managing Terraform and OpenTofu as well. I think it will be good idea to merge `gotofuenv` and [tenv](https://github.com/tofuutils/tenv) together to simplify the future support.

**Reply by dvaumoron** [reply]
gotofuenv is now part of the tofuutils organization (with the managing command renamed `tenv`)

**Comment by rquadling** [comment]
# `terraform-docs`
I've also used [`terraform-docs`](https://github.com/terraform-docs/terraform-docs) a lot (as does probably every public and private module developer). There are no mentions on their website about any sort of support for OpenToFu, no PRs, no issues, [nothing](https://github.com/search?q=repo%3Aterraform-docs%2Fterraform-docs%20tofu&type=code). One of the things about `terraform-docs` is that it uses the [`terraform-config-inspect` library](https://github.com/hashicorp/terraform-config-inspect), which is NOT getting any updates unless under VERY VERY specific conditions set by Hashicorp (see [here](https://github.com/hashicorp/terraform-config-inspect?tab=readme-ov-file#terraform-config-inspect)). As such, a relatively frequent request was to include variable validation in the generated README ([PR#74](https://github.com/hashicorp/terraform-config-inspect/pull/74) and [PR#94](https://github.com/hashicorp/terraform-config-inspect/pull/94) for example.). In the end I made my own [fork](https://github.com/rquadling/terraform-config-inspect) and that allowed me to make a full [PR#675](https://github.com/terraform-docs/terraform-docs/pull/675) in `terraform-docs`. As both the initial issue and the PR are still waiting for replies, this tool may not have full support for OpenToFu.

**Reply by ghost** [reply]
There has been a discussion about this [here](https://github.com/orgs/opentofu/discussions/1018). TL;DR the best course of action would be to reach out to the developers if they would be open to onboarding new maintainers.

**Reply by git-ival** [reply]
In addition to this, it seems that `terraform-docs` doesn't support opentofu's feature for allowing variables in module sources (https://github.com/terraform-docs/terraform-docs/issues/793). So using the tool with a module that utilizes variables in source fields always results in error.

**Reply by siegenthalerroger** [reply]
@rquadling do you have any intention of release builds from your fork? That would be _amazing_

**Reply by rquadling** [reply]
I don't know how to do that yet. And i don't want to pretend my version is official. I just wish I could get my work accepted into the main project. I try to keep it uptodate. The latest tag is https://github.com/rquadling/terraform-docs/releases/tag/v0.21.0-raq.3. This is the official 0.21.0 with my changes.

And the terraform-config-inspect follows the same sort of "thing".

I'm looking to start using `mise` which will allow you to define the versions of tools used in a directory, so those tool versions can be easily maintained if you have MANY repos. I think there's an "include" feature, so keeping things uptodate should also be doable. And `mise` is such that you pretty much do SOMETHING like `go:rquadling/terraform-docs#v0.21.0-raq.3` (something like that - colleagues are using it already and it's 8:30 PM and so I should be doing something else now!).

**Comment by DavidGamba** [comment]
# asdf

Allows you to manage versions of any tool, not just terraform.

There is an OpenTofu plugin already: https://github.com/virtualroot/asdf-opentofu

**Comment by davidjeddy** [comment]
I am big into compliance and static analysis so here are a couple:

[trivy](https://github.com/aquasecurity/trivy)
[kics](https://kics.io/index.html)
[tfsec](https://github.com/aquasecurity/tfsec)

**Comment by virtualroot** [comment]
I created a list of tools that have confirmed OpenTofu support from their maintainers.

https://github.com/virtualroot/awesome-opentofu#tools

**Comment by kvendingoldo** [comment]
# tflint

Discussion thread: https://github.com/aquasecurity/tfsec/issues/2129

**Comment by kvendingoldo** [comment]
Some more tools:

* [digger](https://github.com/diggerhq/digger) - OpenTofu support has been added in scope of [this PR](https://github.com/diggerhq/digger/pull/738)
* [checkov](https://github.com/bridgecrewio/checkov/) - No information about OpenTofu support
* [former2](https://github.com/iann0036/former2) - No information about OpenTofu support
* [infracost](https://github.com/infracost/infracost) - No information about OpenTofu support
* [sato](https://github.com/JamesWoolfenden/sato) - No information about OpenTofu support
* [regula](https://github.com/fugue/regula) - No information about OpenTofu support
* [terraboard](https://github.com/camptocamp/terraboard) - No information about OpenTofu support
* [tfmigrate](https://github.com/minamijoyo/tfmigrate) - OpenTofu support has been added in scope of [this PR](https://github.com/minamijoyo/tfmigrate/issues/162)
* [terrascan](https://github.com/tenable/terrascan) - [Open issue](https://github.com/tenable/terrascan/issues/1609)
* [terraformer](https://github.com/GoogleCloudPlatform/terraformer) - [Open issue](https://github.com/GoogleCloudPlatform/terraformer/issues/1773)
* [pike](https://github.com/JamesWoolfenden/pike) - No information about OpenTofu support
* [terramate](https://github.com/terramate-io/terramate) - OpenTofu support has been added in scope of [this PR](https://github.com/terramate-io/terramate/pull/1306)
* [atlantis](https://github.com/runatlantis/atlantis) - [Open issue](https://github.com/runatlantis/atlantis/issues/3741)
* [terratest](https://github.com/gruntwork-io/terratest) - OpenTofu support has been added in scope of [this PR](https://github.com/gruntwork-io/terratest/issues/1333)
* [terrastate](https://github.com/rohinivsenthil/terrastate) - No information about OpenTofu support
* [terracognita](https://github.com/cycloidio/terracognita) - No information about OpenTofu support
* [inframap](https://github.com/cycloidio/inframap) - No information about OpenTofu support

**Comment by DavidGamba** [comment]
Since the scope of the list is to potentially incorporate some of the tool's ideas into OpenTofu I wrote a tool that allows you to have a config file to:

* automatically inject init backend tfvars and plan tfvars
* automatically load tfvar files based on workspace name 
* allow you to have multiple profiles so that you can have a different config for dev and prod in a single file.
* runs pre-apply checks for you.
* Track the archs you want to use when running terraform provider lock
* Allow you to work on different workspaces for the same project on different terminals

https://github.com/DavidGamba/dgtools/tree/master/bt

**Comment by Ant0wan** [comment]
# tfam
A Rust-based wrapper for concurrent Terraform apply, enabling multi-deployment support.
It is a rewrite in Rust and an enhancement of `tf-autoplan` (not open-sourced).
It works most effectively when used alongside `tfexe` and `tfswitch`.

![tfam interactive demo](https://github.com/Ant0wan/tfam/raw/main/menu.gif)

[https://github.com/Ant0wan/tfam](https://github.com/Ant0wan/tfam)

**Comment by Ant0wan** [comment]
# tfexe
Rust-powered wrapper for seamless execution of tfswitch and Terraform with version control.
It follows this [workflow](https://github.com/Ant0wan/tfexe#how-it-works), making it easy to work with or o customize with any custom binaries.
```mermaid
graph TD
    start(Start)
    TFEXE_set{TFEXE env variable set?}
    TFEXE_unset{TFEXE env variable unset?}

    start --> TFEXE_set
    start --> TFEXE_unset

    TFEXE_set -->|TFEXE value| tfexe_executes[tfexe executes TFEXE]
    TFEXE_unset -->|'tfswitch' is available| tfswitch_executes[tfexe executes tfswitch]
    TFEXE_unset -->|'tfswitch' is not available| terraform_executes[tfexe executes terraform]

    tfexe_executes -->|TFEXE set to 'tofu'| tofu_executes[tfexe executes tofu]
    tfexe_executes -->|TFEXE set to 'terraform'| terraform_executes[tfexe executes terraform]
    tfexe_executes -->|TFEXE set to 'tfswitch'| tfswitch_executes[tfexe executes tfswitch]

    tfswitch_executes -->|execute appropriate 'terraform' version| terraform_executes[tfexe executes terraform]

```

[https://github.com/Ant0wan/tfexe](https://github.com/Ant0wan/tfexe)

**Comment by fergoid** [comment]
# tftest
https://github.com/GoogleCloudPlatform/terraform-python-testing-helper

I have tested this with tofu using tests that inspect both plans and apply outputs and it works seamlessly.  You can pass the binary in when you construct the tftest object.

**Comment by kvendingoldo** [comment]
# tofuenv
https://github.com/tofuutils/tofuenv
[OpenTofu](https://opentofu.org/) version manager inspired by [tfenv](https://github.com/tfutils/tfenv)

**Reply by rquadling** [reply]
How "superceded" is this by [tenv](https://github.com/orgs/opentofu/discussions/1104#discussioncomment-8244166)? If so, is it worth retiring this tool?

**Reply by kvendingoldo** [reply]
@rquadling tenv is successor and I suggest to use tenv instead of tofuenv/tfenv for many reason. First of all, tenv supports Terraform as well as OpenTofu. Secondly, it's a binary file that does not require any dependencies. And at the last, it's more fresh product that we're going to activly develop and support for a long time.

**Comment by kvendingoldo** [comment]
# tenv
https://github.com/tofuutils/tenv
Terraform and OpenTofu version manager, tools written in Go. Successor of [tofuenv](https://github.com/tofuutils/tofuenv) and [tfenv](https://github.com/tfutils/tfenv).

**Comment by x4e-jonas** [comment]
GitLab is [replacing Terraform pipeline](https://gitlab.com/gitlab-org/gitlab/-/issues/440756) templates with OpenTofu: https://gitlab.com/components/opentofu#migrating-from-the-terraform-cicd-templates

However, their [Terraform modules](https://gitlab.com/gitlab-org/gitlab/-/blob/master/lib/gitlab/ci/templates/Terraform-Module.gitlab-ci.yml) template still depends on Terraform.

**Comment by x4e-jonas** [comment]
[vim-terraform](https://hashivim.github.io/vim-terraform/) defaults to `terraform` binary, but can be customized with [`terraform_binary_path`](https://github.com/hashivim/vim-terraform/blob/21f756b933cd11ac5990a6046fdc7c4e2a6c0aee/doc/terraform.txt#L27-L29)

**Comment by diofeher** [comment]
# dependabot
https://github.com/dependabot/dependabot-core

Automatic dependency updates https://github.com/dependabot/dependabot-core/tree/main/opentofu

**Comment by LucasSymons** [comment]
Tag Compliance and automation

[YOR](https://github.com/bridgecrewio/yor) 

This tool is amazing for automating tagging across your stacks. Very easy to config and use. 
Project looks to be in limbo as it transfers owners. 

Works great with `.tf` files, lacks support for `.tofu` files.

**Comment by kinthaiofficial** [comment]
For IaC toolchain composition, a few things worth adding to the list based on what's become standard in production setups:

**Pre-commit hooks**
- `tflint` — catches common HCL mistakes and provider-specific linting before commit
- `terraform-docs` (now tofu-compatible) — auto-generates module documentation from variables/outputs
- `infracost` — cost estimation diff before PR merge; catches accidental expensive resource additions

**Testing layers**
- `tofutest` / `terratest` — integration tests against real cloud resources in ephemeral accounts
- `checkov` or `tfsec` — static security scanning; catches overly permissive IAM, unencrypted storage
- `conftest` with OPA policies — custom policy enforcement beyond what checkov covers

**State management**
- Remote state with locking is table stakes; S3 + DynamoDB or Terraform Cloud equivalent
- State file encryption matters if you're storing sensitive outputs in state
- `tofu state mv` and `tofu import` are frequently needed for drift correction

**Drift detection**
- Running `tofu plan` on a schedule (not just on PR) and alerting on unexpected diff catches manual console changes before they cause incidents

**AI-specific considerations**
For teams using AI agents to generate IaC (increasingly common with coding assistants), you want additional guardrails — schema validation, blast radius estimation before apply, approval gates for changes to stateful resources. The IaC-as-agent-output case is an interesting new challenge: the agent knows syntax but not your org's conventions.

Any specific workflow pain points you're trying to solve with the tool selection?

**Comment by coolapso** [comment]
# tfimport

website: https://tfimport.coolapso.sh/
Github: https://github.com/coolapso/tfimport

Tool to assist with the task of importing infrastructure. Automatically generates `<resourceAddress> <ImportIDs>` for each (supported) provider based on plan or sdk lookups and either generates a import.tf with the import blocks or runs `import <resource> <id>` directly with your IaC Tool of choice (tofu / terraform / terragrunt)

currently supports AWS, Google, Azure, Scaleway, Kubernetes and ArgoCD providers. More providers are relatively easy to add, either by humans or with the help of friendly clanker

**Reply by virtualroot** [reply]
Could you create a PR to include this at https://github.com/virtualroot/awesome-opentofu?
It will be approved and merged immediately :D

**Reply by coolapso** [reply]
on it! where do you thing it fits better? `state` or `helpers` section?

**Reply by virtualroot** [reply]
Let's got with `state` section

**Reply by coolapso** [reply]
done, thanks a lot! :D
